### DISCUSSION TITLE
Can garbage to stdout cause the console's input mode to change?

### DISCUSSION THREAD

**Discussion by vefatica** [microsoft/terminal#15685] [General]
TCC 30.00.22 x64
Microsoft Windows 10 Pro for Workstations
10.0.19045.3086 (2009, 22H2)
WindowsTerminal 1.18.2305.26002 (preview)

Using Cygwin's CAT.EXE ((GNU coreutils) 9.0), if I CAT some binary file (by mistake the first time) the console's input mode is very likely to be changed to include ENABLE_VIRTUAL_TERMINAL_INPUT.  Whether or not that happens depends on the binary file and is consistent for a given file. This has a devastating effect on my shell (TCC) but it also happens in PowerShell/PWSH (apparently without the devastating effect).  For example, in PowerShell, before and after issuing
 `d:\cygwin\bin\cat.exe d:\wt\exe\WindowsTerminal.exe` ... before ...

```
V:\> conmode /i

Console input mode flags (0x1F7)

    ENABLE_PROCESSED_INPUT (0x1)
    ENABLE_LINE_INPUT (0x2)
    ENABLE_ECHO_INPUT (0x4)
    ENABLE_MOUSE_INPUT (0x10)
    ENABLE_INSERT_MODE (0x20)
    ENABLE_QUICK_EDIT_MODE (0x40)
    ENABLE_EXTENDED_FLAGS (0x80)
    ENABLE_AUTO_POSITION (0x100)
```

after ...

```
V:\> conmode /i

Console input mode flags (0x2E8)

    ENABLE_WINDOW_INPUT (0x8)
    ENABLE_INSERT_MODE (0x20)
    ENABLE_QUICK_EDIT_MODE (0x40)
    ENABLE_EXTENDED_FLAGS (0x80)
    ENABLE_VIRTUAL_TERMINAL_INPUT (0x200)
```

The same happens in a Windows console.  It does not happen if I use another means of sending the file to stdout ... TCC's TYPE, PowerShell's TYPE, an older (32-bit) Gnu CAT.EXE ((GNU coreutils) 5.3.0), WSL CAT (each of those producing visually different output).

I wouldn't expect Cygwin's CAT.EXE to do that on purpose.  And whether or not it happens depends on the binary file.  So I must ask whether sending garbage to stdout could cause it.

**Comment by DHowett** [comment]
Excellent question! The trick with anything running under Cygwin (and MSYS2 by extension) is that _it is not dumping text to stdout._ One of its default modes of operation simulates a terminal using the Win32 console APIs by processing stdout for control sequences.

And there are absolutely control sequences that change how input is generated.

It's not a stretch to imagine that Cygwin is detecting one of those control sequences and translating it _into_ a request for `SetConsoleMode`. The bug is moreso that it is not restoring the right modes when it exits.

**Comment by vefatica** [comment]
OK, thanks!  Do I have any choice regarding Cygwin's default MO?

I only got Cygwin to get 64-bit versions of my old GnuWin32 tools.  To be honest, in normal use, they don't make any difference.  And CAT is one I almost never use.

**Reply by DHowett** [reply]
Hmmm. I'd be interested in seeing what `$TERM` is set to inside Cygwin. As of version 3.1 or 3.2, they added support for the Windows 10 console. If that support is being used, `TERM` will be `xterm-256color`. If _not_ (and they are using translation,) `TERM` will be `cygwin`.

**Comment by vefatica** [comment]
If I just start d:\cygwin\bin\cat.exe (it'll wait to process stdin), TERM is not in it's Windows environment.  If I set TERM in my shell before starting CAT.EXE, it's inherited, but using either of the values you mentioned doesn't alter the behavior in question. [Was that test meaningful?]

**Comment by ClaireCJS** [comment]
This bug has plagued me for years, but I keep stubbornly using cat.exe instead of the internal type command anyway.

The end-result is, tab filename completion no longer works, and the shell has to be terminated and re-opened entirely. It's awful.

**Comment by vefatica** [comment]
OT:  What is ENABLE_AUTO_POSITION (0x100)?  I see it in OpenConsole and not in Conhost.  What does it do?

**Reply by zadjii-msft** [reply]
Interesting. It's not really used for anything in the current codebase. 

https://github.com/search?q=repo%3Amicrosoft%2Fterminal%20ENABLE_AUTO_POSITION&type=code

This would require much more historical digging to figure out what it _was_.

**Reply by DHowett** [reply]
> not really used for anything

It's translated into `CONSOLE_AUTO_POSITION`, which _is_ used.

Promoted:

https://github.com/microsoft/terminal/blob/2f0d3dc17a09493a17ef45551deea864914854c2/src/host/getset.cpp#L56

Used:

https://github.com/microsoft/terminal/blob/2f0d3dc17a09493a17ef45551deea864914854c2/src/interactivity/win32/menu.cpp#L323

https://github.com/microsoft/terminal/blob/2f0d3dc17a09493a17ef45551deea864914854c2/src/interactivity/win32/window.cpp#L267-L301

**Reply by zadjii-msft** [reply]
> `ENABLE_AUTO_POSITION`

> `CONSOLE_AUTO_POSITION`

ngl I didn't even see that those were different. I need more sleep 😪
